test/unit/spec/twilioconnection.js

```diff
@@ -173,185 +234,294 @@ describe('TwilioConnection', function() {
+                      assert.equal(wsCloseCallCount, keepAlive ? 0 : 1);
+                    });
+
+                    if (keepAlive) {
+                      it('should eventually transition to "connecting" state and send a "hello" message with the requested heartbeat timeout', async () => {
+                        changedState = await new Promise(resolve => twilioConnection.once('stateChanged', resolve));
+                        assert.equal(changedState, 'connecting');
+                        assert.equal(twilioConnection.state, 'connecting');
+                        const hello = JSON.parse(twilioConnection._ws.send.args[1][0]);
+                        assert.equal(typeof hello.id, 'string');
+                        assert.equal(hello.timeout, twilioConnection._options.requestedHeartbeatTimeout);
+                        assert.equal(hello.type, 'hello');
+                        assert.equal(hello.cookie, 'foo');
+                        if (options.helloBody) {
+                          assert.equal(hello.body, options.helloBody);
+                        } else {
+                          assert.equal('body' in hello, false);
+                        }
+                      });
+                    } else {
+                      context('should eventually transition to "early" state and when a new WebSocket is opened', () => {
+                        beforeEach(async () => {
+                          changedState = await new Promise(resolve => twilioConnection.once('stateChanged', resolve));
+                          assert.equal(changedState, 'early');
+                          assert.equal(twilioConnection.state, 'early');
+                          const stateChanged = new Promise(resolve => twilioConnection.once('stateChanged', resolve));
+                          twilioConnection._ws.open();
+                          changedState = await stateChanged;
+                        });
+
+                        it('should eventually transition to "connecting" state and send a "hello" message with the requested heartbeat timeout', () => {
+                          assert.equal(changedState, 'connecting');
+                          assert.equal(twilioConnection.state, 'connecting');
+                          const hello = JSON.parse(twilioConnection._ws.send.args[0][0]);
+                          assert.equal(typeof hello.id, 'string');
+                          assert.equal(hello.timeout, twilioConnection._options.requestedHeartbeatTimeout);
+                          assert.equal(hello.type, 'hello');
+                          assert.equal(hello.cookie, 'foo');
+                          if (options.helloBody) {
+                            assert.equal(hello.body, options.helloBody);
+                          } else {
+                            assert.equal('body' in hello, false);
+                          }
+                        });
+                      });
```

Comment:
Is it possible to test that retryAfter is respected rather than a hello being eventually sent?

Note: The cookie should be optional, if the server busy does not include one then the client hello should not either.